### PR TITLE
Pr for issue 751

### PR DESCRIPTION
**Pull request by rebcabin**
To simplify the PR, I removed all the application files "GS*.m". Most of the commits pertain to those file. I just wanted to leave in the commits that fixed bugs. 

I will be happy to contribute those application files at some future time, when I've finished them. Most of them imply some very useful unit tests.

### CODE REVIEW

**Comment by poeschko** [comment]
Perhaps not related to this pull request, but could you take a look at the failing Travis CI build?

**Comment by GarkGarcia** [comment]
Hey @rebcabin, would you be willing to git-rebase the PR branch?

File: mathics/builtin/lists.py [hunk 1]
```diff
@@ -1734,15 +1734,18 @@ class _IterationFunction(Builtin):
     def get_result(self, items):
         pass
 
-    def apply_range(self, expr, i, imax, evaluation):
+    def apply_imax_or_list(self, expr, i, imax, evaluation):
```
Comment (GarkGarcia) [inline mathics/builtin/lists.py:1737]: This is completely unrelated to the PR.

File: mathics/core/pattern.py [hunk 2]
```diff
@@ -289,27 +289,28 @@ def per_expr(yield_expr, expr_groups, sum=0):
 
                         if expr_groups:
                             expr, count = expr_groups.popitem()
-                            max_per_pattern = count / len(patterns)
+                            max_per_pattern = int(count / len(patterns))
```
Comment (GarkGarcia) [inline mathics/core/pattern.py:292]: This is not related too.

File: mathics/packages/Issue748.m [hunk 3]
```diff
@@ -0,0 +1,30 @@
+ClearAll[parse]
```
Comment (GarkGarcia) [inline mathics/packages/Issue748.m:1]: This doesn't seem to be related too.

**Comment by GarkGarcia** [comment]
See https://github.com/mathics/Mathics/pull/838 for updates.
